bentoml/server/middlewares.py

```diff
@@ -0,0 +1,66 @@
+from timeit import default_timer
+
+from flask import Request
+
+from bentoml import config
+
+
+class InstrumentMiddleware:
+    def __init__(self, app, bento_service):
+        self.app = app
+        self.bento_service = bento_service
+
+        from prometheus_client import Histogram, Counter, Gauge
+
+        service_name = self.bento_service.name
+        namespace = config('instrument').get('default_namespace')
+
+        self.metrics_request_duration = Histogram(
+            name=service_name + '_request_duration_seconds',
+            documentation=service_name + " API HTTP request duration in seconds",
+            namespace=namespace,
+            labelnames=['api_name', 'service_version', 'http_response_code'],
+        )
+        self.metrics_request_total = Counter(
+            name=service_name + "_request_total",
+            documentation='Totoal number of HTTP requests',
+            namespace=namespace,
+            labelnames=['api_name', 'service_version', 'http_response_code'],
+        )
+        self.metrics_request_in_progress = Gauge(
+            name=service_name + "_request_in_progress",
+            documentation='Totoal number of HTTP requests in progress now',
+            namespace=namespace,
+            labelnames=['api_name', 'service_version'],
+        )
+
+    def __call__(self, environ, start_response):
+        req = Request(environ)
+        api_name = req.path.lstrip('/')
+        start_time = default_timer()
+
+        def start_response_wrapper(status, headers):
```

Comment:
start_response callable got called at the time of creating a response right?  why not just wrap those code the same way `metrics_request_in_progress ` is done below, wrapping around `self.app` call?

Reply:
Right.
The reason here is to get the response status code.